dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/orgunit/OrgUnitMergeRequest.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.merge.orgunit;
+
+import static org.apache.commons.lang3.ObjectUtils.firstNonNull;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import lombok.Getter;
+
+import org.hisp.dhis.common.IdentifiableObjectUtils;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * Encapsulation of an org unit merge request.
+ *
+ * @author Lars Helge Overland
+ */
+@Getter
+public class OrgUnitMergeRequest
+{
+    private Set<OrganisationUnit> sources = new HashSet<>();
+
+    private OrganisationUnit target;
+
+    private DataMergeStrategy dataValueMergeStrategy;
+
+    private DataMergeStrategy dataApprovalMergeStrategy;
+
+    private boolean deleteSources;
+
+    public Set<OrganisationUnit> getSources()
+    {
+        return ImmutableSet.copyOf( sources );
+    }
+
+    public OrganisationUnit getTarget()
+    {
+        return target;
+    }
+
+    @Override
+    public String toString()
+    {
+        return MoreObjects.toStringHelper( this )
+            .add( "sources", IdentifiableObjectUtils.getIdentifiers( sources ) )
+            .add( "target", target != null ? target.getId() : null )
+            .add( "deleteSources", deleteSources )
+            .toString();
+    }
+
+    public static class Builder
```

Comment:
A bit odd "not really immutable" builder pattern here - you could use lombok generated `builder()` method and provide the utilities methods like `addSource` directly on `OrgUnitMergeRequest` and implement it using `builder().sources(newList).build()` and alike.

Reply:
Okay, I used a custom builder to set default values and have a custom add source method.

How is this builder not immutable?

Reply:
The `OrgUnitMergeRequest` isn't immutable as fields are not final. I could build the request using a builder pass it on and then modify it using the same builder instance.

Reply:
Yeah agree we should have made a copy/new instance of the request object in the build method.